core/sail/inferencer/src/main/java/org/eclipse/rdf4j/sail/inferencer/fc/ForwardChainingSchemaCachingRDFSInferencer.java

```diff
@@ -0,0 +1,318 @@
+    Map<Resource, Set<Resource>> calculatedDomain = new HashMap<>();
+
+    // The inferencer has been instantiated from another inferencer and shares it's schema with that one
+    private boolean sharedSchema;
+
+    void clearInferenceTables() {
+        properties = new HashSet<>();
+        types = new HashSet<>();
+        subClassOfStatements = new ArrayList<>();
+        subPropertyOfStatements = new ArrayList<>();
+        rangeStatements = new ArrayList<>();
+        domainStatements = new ArrayList<>();
+
+        calculatedTypes = new HashMap<>();
+        calculatedProperties = new HashMap<>();
+        calculatedRange = new HashMap<>();
+        calculatedDomain = new HashMap<>();
+    }
+
+    void readLock(ForwardChainingSchemaCachingRDFSInferencerConnection connection) {
+
+//        if (numberOfThreadsWaitingForWriteLock.get() > 0) {
+//            System.err.println("starve reads");
+//        }
+
+        while (numberOfThreadsWaitingForWriteLock.get() > 0) {
+            Thread.yield();
+        }
+
+        if (connection.lockStamp != 0) {
+            throw new SailException("Connection already has a lock! Might be a begin() without a commit() or rollback() for previous transaction().");
+        }
+        connection.lockStamp = readWriteLock.readLock();
+//        System.err.println("readLock: " + connection.lockStamp);
+
+    }
+
+    void releaseLock(ForwardChainingSchemaCachingRDFSInferencerConnection connection) {
+        if(connection.lockStamp == 0){
+            throw new SailException("Expected connection to have lock. Might be a commit() without a begin().");
+        }
+        readWriteLock.unlock(connection.lockStamp);
+//        System.err.println("Released lock: " + connection.lockStamp);
+
+        connection.lockStamp = 0;
+    }
+
+
+    private AtomicInteger numberOfThreadsWaitingForWriteLock = new AtomicInteger(0);
+
+    void upgradeLock(ForwardChainingSchemaCachingRDFSInferencerConnection connection) {
+
+//        System.err.println("Attempt writelock: "+connection.lockStamp);
+
+        numberOfThreadsWaitingForWriteLock.incrementAndGet();
+
+        try {
+           while(true) {
+                long l = readWriteLock.tryConvertToWriteLock(connection.lockStamp);
+
+                if (l != 0) {
+//                    long temp = connection.lockStamp;
+                    connection.lockStamp = l;
+//                    if (temp != l) {
+//                        System.err.println("readLock: " + temp + " writeLock: " + connection.lockStamp);
+//                    }
+
+
+                    return;
+                }
+
+                try {
+                    Thread.sleep(random.nextInt(2));
+                } catch (InterruptedException e) {
+                    // ignore interrupted exception
+                }
+
+                // detect potential deadlock scenario
+                if(numberOfThreadsWaitingForWriteLock.get() > 1  //More than 1 connection waiting for a write lock
```

Comment:
Cannot rely on two successive calls to numberOfThreadsWaitingForWriteLock.get() to return the same value. In order to reason about your custom locking procedure, you should at least cache the value before the if statement and use it internally.